### DISCUSSION TITLE
some issues about subword-level nmt.

### DISCUSSION THREAD

**Discussion by 4hfly** [4hfly/Associated_Learning#20] [General]
- codes 都在 `branch: nmt`: `./nmt/subword_level/`

## 首先是原來架構的疑問
1. embedding layer 是否需要 training？我覺得好像沒有必要去計算 mse loss，直接過 `nn.Embedding()` 就好了。
2. 現在訓練時是用 mse + cos，inference 是用根據 vocab 做 cross entropy，所以之後都改用 mse + cross entropy？
3. 本來有個 vocab.py，把裡面 function 抽出來用的原因是（像是`to_input_tensor`)？

## 然後說一下改動的部分
1. 我先不額外 call LSTMAL，直接寫兩層 layer 在 nmt.py 裡，我想這樣比較好懂。
2. spm 在 utils.py 的 `read_corpus` 裡就會 encode 成 str 了，之後才會轉成 id，跟原版的做法一樣。
3. 有寫東西的幾乎都在 nmt.py，其他可以忽略。
4. `TODO`下的 code 是之後要寫回去 model.py 的。

## 其他
因為我不曉得你後來用的 nmt 是哪一份，所以你有其他完成的部分可以再讓我看看，
我這份有另外去看 cs224n 的作業說明，變數名稱有改成我比較能理解的命名方式。

可以的話最好把 almodel.py, almodel2.py, nmt_new.py 這些整理一下，有 git 不用怕刪了會找不回來，每個 .py 留一份最新的就好。

應該還有其他要討論的，想到會再補充。
@Hibb-bb

**Comment by Hibb-bb** [comment]
section1:
  1. embedding 可以不用那些loss
  2. training 的時候還是有用到cross entropy 從embedding -> readout(linear)那邊
  3. yes

I will read your code tomorrow, you can start to train text classification tasks first, you can use IMDB or another dataset, I think it can run on prof. Chen's lab server also.

I'm currently modifying the code to original forward, so I might need you to train text classification first to report progress next Monday. I think it might be really similar to normal nmt architecture, but im still working on it.
